scene/resources/visual_shader_particle_nodes.cpp

```diff
@@ -1162,6 +1162,206 @@ VisualShaderNodeParticleAccelerator::VisualShaderNodeParticleAccelerator() {
+	vec4 result_4567 = vec4(
+		dot(delta - vec4(0.0, 0.0, 1.0, 0.0), grad_0010), dot(delta - vec4(1.0, 0.0, 1.0, 0.0), grad_1010),
+		dot(delta - vec4(0.0, 1.0, 1.0, 0.0), grad_0110), dot(delta - vec4(1.0, 1.0, 1.0, 0.0), grad_1110));
+	vec4 result_89AB = vec4(
+		dot(delta - vec4(0.0, 0.0, 0.0, 1.0), grad_0001), dot(delta - vec4(1.0, 0.0, 0.0, 1.0), grad_1001),
+		dot(delta - vec4(0.0, 1.0, 0.0, 1.0), grad_0101), dot(delta - vec4(1.0, 1.0, 0.0, 1.0), grad_1101));
+	vec4 result_CDEF = vec4(
+		dot(delta - vec4(0.0, 0.0, 1.0, 1.0), grad_0011), dot(delta - vec4(1.0, 0.0, 1.0, 1.0), grad_1011),
+		dot(delta - vec4(0.0, 1.0, 1.0, 1.0), grad_0111), dot(delta - vec4(1.0, 1.0, 1.0, 1.0), grad_1111));
+	vec4 fade = delta * delta * delta * (10.0 + delta * (-15.0 + delta * 6.0));
+	vec4 result_W0 = mix(result_0123, result_89AB, fade.w), result_W1 = mix(result_4567, result_CDEF, fade.w);
+	vec4 result_WZ = mix(result_W0, result_W1, fade.z);
+	vec2 result_WZY = mix(result_WZ.xy, result_WZ.zw, fade.y);
+	return mix(result_WZY.x, result_WZY.y, fade.x);
+}
+
+// Curl 3D and three-noise function with friendly permission by Isaac Cohen.
+// Modified to accept 4D noise.
+vec3 turb_noise_3x(vec4 p) {
+	float s = turb_noise(p);
+	float s1 = turb_noise(p + vec4(vec3(0.0), 1.7320508 * 2048.333333));
+	float s2 = turb_noise(p - vec4(vec3(0.0), 1.7320508 * 2048.333333));
+	vec3 c = vec3(s, s1, s2);
+	return c;
+}
+
+vec3 turb_curl_3d(vec4 p, float c) {
+	float epsilon = 0.001 + c;
+	vec4 dx = vec4(epsilon, 0.0, 0.0, 0.0);
+	vec4 dy = vec4(0.0, epsilon, 0.0, 0.0);
+	vec4 dz = vec4(0.0, 0.0, epsilon, 0.0);
+	vec3 x0 = turb_noise_3x(p - dx).xyz;
+	vec3 x1 = turb_noise_3x(p + dx).xyz;
+	vec3 y0 = turb_noise_3x(p - dy).xyz;
+	vec3 y1 = turb_noise_3x(p + dy).xyz;
+	vec3 z0 = turb_noise_3x(p - dz).xyz;
+	vec3 z1 = turb_noise_3x(p + dz).xyz;
+	float x = (y1.z - y0.z) - (z1.y - z0.y);
+	float y = (z1.x - z0.x) - (x1.z - x0.z);
+	float z = (x1.y - x0.y) - (y1.x - y0.x);
+	return normalize(vec3(x, y, z));
+}
+
+vec3 get_turb_noise_direction(vec3 __pos, float __turb_n_strength, float __turb_n_scale, vec3 __turb_n_speed, float __turb_n_speed_random) {
+	float __adj_contrast = max((__turb_n_strength - 1.0), 0.0) * 70.0;
+	vec4 __noise_time = TIME * vec4(__turb_n_speed, __turb_n_speed_random);
+	vec4 __noise_pos = vec4(__pos * __turb_n_scale, 0.0);
+	vec3 __noise_direction = turb_curl_3d(__noise_pos + __noise_time, __adj_contrast);
+	__noise_direction = mix(0.9 * __noise_direction, __noise_direction, __turb_n_strength - 9.0);
+	return __noise_direction;
+}
+
+float get_turb_influence(uint __noise_seed, float __turb_influence_min, float __turb_influence_max, float __turb_influence_over_life) {
+	float __turb_influence_range = mix(__turb_influence_min, __turb_influence_max, __rand_from_seed(__noise_seed));
+	float __turb_influence = clamp(__turb_influence_range * __turb_influence_over_life, 0.0, 1.0);
+	return __turb_influence;
+}
+	)");
+
+	return code;
```

Comment:
Do you think it would be feasible/make sense to share this with `ParticleProcessMaterial` to avoid duplication?

Reply:
I don't think it makes sense here to combine these because I had to change quite a few parts of the shader code to make it usable for visual shaders due to the separation of the `start` and `process` functions in the particle shader. Furthermore, the `ParticleProcessMaterial` can't be converted to a visual shader anyway so I currently see no reason to do so.